File: backend/src/services/auth.service.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const prisma = require('../lib/prisma');
const config = require('../config');
const ApiError = require('../utils/ApiError');

/**
 * Helper: Serialize BigInt fields untuk JSON response
 */
const serializeUser = (user) => ({
  ...user,
  id: user.id.toString(),
});

/**
 * Register user baru
 */
const register = async ({ name, email, password, phoneNumber, address }) => {
  // Cek email sudah terpakai
  const existingUser = await prisma.user.findUnique({ where: { email } });
  if (existingUser) {
    throw ApiError.conflict('Email sudah terdaftar.');
  }

  // Cek phone number jika diberikan
  if (phoneNumber) {
    const existingPhone = await prisma.user.findUnique({
      where: { phoneNumber },
    });
    if (existingPhone) {
      throw ApiError.conflict('Nomor HP sudah terdaftar.');
    }
  }

  // Hash password
  const passwordHash = await bcrypt.hash(password, 12);

  // Create user
  const user = await prisma.user.create({
    data: {
      name,
      email,
      phoneNumber: phoneNumber || null,
      passwordHash,
      address: address || null,
    },
    select: {
      id: true,
      name: true,
      email: true,
      phoneNumber: true,
      profilePhotoUrl: true,
      address: true,
      createdAt: true,
    },
  });

  // Generate JWT
  const token = generateToken(user.id);

  return { user: serializeUser(user), token };
};

/**
 * Login user
 */
const login = async ({ email, password }) => {
  // Cari user by email
  const user = await prisma.user.findUnique({
    where: { email },
  });

  if (!user) {
    throw ApiError.unauthorized('Email atau password salah.');
  }

  if (!user.isActive) {
    throw ApiError.unauthorized('Akun telah dinonaktifkan.');
  }

  // Verifikasi password
  const isPasswordValid = await bcrypt.compare(password, user.passwordHash);
  if (!isPasswordValid) {
    throw ApiError.unauthorized('Email atau password salah.');
  }

  // Generate JWT
  const token = generateToken(user.id);

  const userData = {
    id: user.id,
    name: user.name,
    email: user.email,
    phoneNumber: user.phoneNumber,
    profilePhotoUrl: user.profilePhotoUrl,
    address: user.address,
  };

  return { user: serializeUser(userData), token };
};

/**
 * Get current user profile (dari token)
 */
const getMe = async (userId) => {
  const user = await prisma.user.findUnique({
    where: { id: BigInt(userId) },
    select: {
      id: true,
      name: true,
      email: true,
      phoneNumber: true,
      profilePhotoUrl: true,
      address: true,
      emailVerifiedAt: true,
      isActive: true,
      createdAt: true,
      updatedAt: true,
    },
  });

  if (!user) {
    throw ApiError.notFound('User tidak ditemukan.');
  }

  return serializeUser(user);
};

/**
 * Generate JWT token
 */
const generateToken = (userId) => {
  return jwt.sign(
    { id: userId.toString() },
    config.jwt.secret,
    { expiresIn: config.jwt.expiresIn }
  );
};

module.exports = {
  register,
  login,
  getMe,
};

File: frontend/lib/services/auth_service.dart
import 'api_service.dart';

class AuthService {
  /// Login — returns { user, token }
  static Future<Map<String, dynamic>> login(String email, String password) async {
    final response = await ApiService.post('/auth/login', body: {
      'email': email,
      'password': password,
    });
    final data = response['data'];
    // Save token
    await ApiService.saveToken(data['token']);
    return data;
  }

  /// Register — returns { user, token }
  static Future<Map<String, dynamic>> register({
    required String name,
    required String email,
    required String password,
    String? address,
    String? phoneNumber,
    DateTime? birthDate,
  }) async {
    final body = <String, dynamic>{
      'name': name,
      'email': email,
      'password': password,
    };
    if (address != null && address.isNotEmpty) body['address'] = address;
    if (phoneNumber != null && phoneNumber.isNotEmpty) body['phoneNumber'] = phoneNumber;
    if (birthDate != null) body['birthDate'] = birthDate.toIso8601String().split('T')[0];

    final response = await ApiService.post('/auth/register', body: body);
    final data = response['data'];
    // Save token
    await ApiService.saveToken(data['token']);
    return data;
  }

  /// Get current user profile
  static Future<Map<String, dynamic>> getMe() async {
    final response = await ApiService.get('/auth/me');
    return Map<String, dynamic>.from(response['data']);
  }

  /// Logout — clear token
  static Future<void> logout() async {
    await ApiService.clearToken();
  }
}
